src/CoSELayout.js

```diff
@@ -1425,8 +1425,129 @@ CoSELayout.prototype.tileCompoundMembers = function (childGraphMap, idToNode) {
+  }
+
+  // return the normalized ratio
+  return ratio;
+};
+
+/*
+ * Calculates the ideal width for the rows. This method assumes that
+ * each node has the same sizes and calculates the ideal row width that
+ * approximates a square shaped complex accordingly. However, since nodes would
+ * have different sizes some rows would have different sizes and the resulting
+ * shape would not be an exact square.
+ */
+CoSELayout.prototype.calcIdealRowWidth = function (members, favorHorizontalDim) {
+  // To approximate a square shaped complex we need to make complex width equal to complex height.
+  // To achieve this we need to solve the following equation system for hc:
+  // (x + bx) * hc - bx = (y + by) * vc - by, hc * vc = n
+  // where x is the avarage width of the nodes, y is the avarage height of nodes
+  // bx and by are the buffer sizes in horizontal and vertical dimensions accordingly,
+  // hc and vc are the number of rows in horizontal and vertical dimensions
+  // n is number of members.
+
+  var verticalPadding = CoSEConstants.TILING_PADDING_VERTICAL;
+  var horizontalPadding = CoSEConstants.TILING_PADDING_HORIZONTAL;
+
+  // number of members
+  var membersSize = members.length;
+
+  // sum of the width of all members
+  var totalWidth = 0;
+
+  // sum of the height of all members
+  var totalHeight = 0;
+
+  var maxWidth = 0;
+
+  // traverse all members to calculate total width and total height and get the maximum members width
+  members.forEach(function (node) {
+    totalWidth += node.getWidth();
+    totalHeight += node.getHeight();
+
+    if (node.getWidth() > maxWidth) {
+      maxWidth = node.getWidth();
+    }
+  });
+
+  // average width of the members
+  var averageWidth = totalWidth / membersSize;
+
+  // average height of the members
+  var averageHeight = totalHeight / membersSize;
+
+  // solving the initial equation system for the hc yields the following second degree equation:
+  // hc^2 * (x+bx) + hc * (by - bx) - n * (y + by) = 0
+
+  // the delta value to solve the equation above for hc
+  var delta = Math.pow(verticalPadding - horizontalPadding, 2) + 4 * (averageWidth + horizontalPadding) * (averageHeight + verticalPadding) * membersSize;
+
+  // solve the equation using delta value to calculate the horizontal count
+  // that represents the number of nodes in an ideal row
+  var horizontalCountDouble = (horizontalPadding - verticalPadding + Math.sqrt(delta)) / (2 * (averageWidth + horizontalPadding));
+  // round the calculated horizontal count up or down according to the favored dimension
+  var horizontalCount;
+
+  if (favorHorizontalDim) {
+    horizontalCount = Math.ceil(horizontalCountDouble);
+    // if horizontalCount count is not a float value then both of rounding to floor and ceil
+    // will yield the same values. Instead of repeating the same calculation try going up
+    // while favoring horizontal dimension in such cases
+    if ( horizontalCount == horizontalCountDouble ) {
```

Comment:
Here, let's say `horizontalCountDouble` is equal to 4. Then `horizontalCount` will also be 4. Since they are equal, we make `horizontalCount` 5 in below line. Is this what we want or do we want to keep it as 4 since it's not a float?

Reply:
I got it, no problem.